File: dron_controller_c_plus_plus/mpu_test.cpp
#include "tests.h"

int mpu_basic_test() {
    MPU6050 mpu(0x68);  // Initialize MPU6050 with address 0x68

    mpu.reset_mpu6050();
    std::cout << "Temperature: " << mpu.get_temp() << " °C" << std::endl;

    double accel_data[3];
    uint8_t accel_range=mpu.read_accel_range();
    mpu.get_accel_data(accel_data,accel_range);
    std::cout << "Accelerometer data:" << std::endl;
    std::cout << "X: " << accel_data[0] << std::endl;
    std::cout << "Y: " << accel_data[1] << std::endl;
    std::cout << "Z: " << accel_data[2] << std::endl;

    double gyro_data[3];
    uint8_t gyro_range = mpu.read_gyro_range();
    mpu.get_gyro_data(gyro_data,gyro_range);
    std::cout << "Gyroscope data:" << std::endl;
    std::cout << "X: " << gyro_data[0] << std::endl;
    std::cout << "Y: " << gyro_data[1] << std::endl;
    std::cout << "Z: " << gyro_data[2] << std::endl;

    mpu.sleep(); // Set MPU6050 to sleep

    return 0;
}


void calibrate_mpu6050(MPU6050& mpu, double* acc_bias, double* gyroscope_bias, int duration = 5) {
    int num_samples = duration * 100;  // Taking 100 samples per second
    double accel_bias[3] = {0.0, 0.0, 0.0};
    double gyro_bias[3] = {0.0, 0.0, 0.0};
    uint8_t gyro_range = mpu.read_gyro_range();
    uint8_t accel_range=mpu.read_accel_range();

    for (int i = 0; i < num_samples; ++i) {
        double accel_data[3];
        double gyro_data[3];

        mpu.get_accel_data(accel_data,accel_range, false);
        mpu.get_gyro_data(gyro_data,gyro_range);

        accel_bias[0] += accel_data[0];
        accel_bias[1] += accel_data[1];
        accel_bias[2] += accel_data[2];

        gyro_bias[0] += gyro_data[0];
        gyro_bias[1] += gyro_data[1];
        gyro_bias[2] += gyro_data[2];

        usleep(10000); // Sleep for 10 000 microseconds between samples
    }

    accel_bias[0] /= num_samples;
    accel_bias[1] /= num_samples;
    accel_bias[2] /= num_samples;
    gyro_bias[0] /= num_samples;
    gyro_bias[1] /= num_samples;
    gyro_bias[2] /= num_samples;

    // Adjust the accelerometer bias to account for gravity on the Z-axis
    accel_bias[2] -= MPU6050::GRAVITIY_MS2;

    if (acc_bias != nullptr) {
        acc_bias[0] = accel_bias[0];
        acc_bias[1] = accel_bias[1];
        acc_bias[2] = accel_bias[2];
    }
    if (gyroscope_bias != nullptr) {
        gyroscope_bias[0] = gyro_bias[0];
        gyroscope_bias[1] = gyro_bias[1];
        gyroscope_bias[2] = gyro_bias[2];
    }

    // Print biases for debugging
    std::cout << "Accelerometer Bias: x=" << accel_bias[0] << ", y=" << accel_bias[1] << ", z=" << accel_bias[2] << std::endl;
    std::cout << "Gyroscope Bias: x=" << gyro_bias[0] << ", y=" << gyro_bias[1] << ", z=" << gyro_bias[2] << std::endl;
}

void calculate_angles(MPU6050& mpu, double* acc_bias, double* gyro_bias) {

    const double DT = 0.07;
    const double AA = 0.98;
    double gyroXangle = 0.0, gyroYangle = 0.0, gyroZangle = 0.0;
    double CFangleX = 0.0, CFangleY = 0.0;

    uint8_t gyro_range = mpu.read_gyro_range();
    uint8_t accel_range=mpu.read_accel_range();

    int cnt = 0;
    while (true) {
        cnt++;
        auto t1 = std::chrono::steady_clock::now();

        double accel_data[3];
        double gyro_data[3];

        mpu.get_accel_data(accel_data,accel_range, false);
        mpu.get_gyro_data(gyro_data,gyro_range);

        gyro_data[0] -= gyro_bias[0];
        gyro_data[1] -= gyro_bias[1];
        gyro_data[2] -= gyro_bias[2];

        accel_data[0] -= acc_bias[0];
        accel_data[1] -= acc_bias[1];
        accel_data[2] -= acc_bias[2];

        // Calculate angles from accelerometer
        double AccXangle = atan2(accel_data[0], sqrt(accel_data[1] * accel_data[1] + accel_data[2] * accel_data[2])) * 180.0 / M_PI;
        double AccYangle = atan2(accel_data[1], accel_data[2]) * 180.0 / M_PI;

        // Calculate angles from gyroscope
        gyroXangle += gyro_data[0] * DT;
        gyroYangle += gyro_data[1] * DT;
        gyroZangle += gyro_data[2] * DT;

        // Apply complementary filter
        CFangleX = AA * (CFangleX + gyro_data[0] * DT) + (1 - AA) * AccXangle;
        CFangleY = AA * (CFangleY + gyro_data[1] * DT) + (1 - AA) * AccYangle;

        if (cnt == 20) {
            std::cout << "Pitch (CFangleX): " << CFangleX << " degrees, Roll (CFangleY): " << CFangleY << " degrees" << std::endl;
            cnt = 0;
        }

        auto t2 = std::chrono::steady_clock::now();
        //duration is in seconds
        usleep(1000000*std::max(0.0, DT - std::chrono::duration<double>(t2 - t1).count()));
    }
}

File: dron_controller/mpu_test.py
import time
from mpu6050 import mpu6050
import math


def calibrate_mpu6050(mpu, duration=5):
    """
    Calibrate the accelerometer and gyroscope biases.

    :param mpu: The mpu6050 object.
    :param duration: The duration in seconds to average the readings over.
    :return: A dictionary with the calculated biases for accelerometer and gyroscope.
    """
    num_samples = duration * 100  # Taking 100 samples per second
    accel_bias = {'x': 0.0, 'y': 0.0, 'z': 0.0}
    gyro_bias = {'x': 0.0, 'y': 0.0, 'z': 0.0}

    gyro_range = mpu.read_gyro_range(True)
    accel_range = mpu.read_accel_range(True)

    for _ in range(num_samples):
        accel_data = mpu.get_accel_data(accel_range)
        gyro_data = mpu.get_gyro_data(gyro_range)

        accel_bias['x'] += accel_data['x']
        accel_bias['y'] += accel_data['y']
        accel_bias['z'] += accel_data['z']

        gyro_bias['x'] += gyro_data['x']
        gyro_bias['y'] += gyro_data['y']
        gyro_bias['z'] += gyro_data['z']

        time.sleep(0.01)  # Sleep for 10 milliseconds between samples

    accel_bias['x'] /= num_samples
    accel_bias['y'] /= num_samples
    accel_bias['z'] /= num_samples
    gyro_bias['x'] /= num_samples
    gyro_bias['y'] /= num_samples
    gyro_bias['z'] /= num_samples

    # Adjust the accelerometer bias to account for gravity on the Z-axis
    accel_bias['z'] -= mpu.GRAVITIY_MS2

    return {'accel_bias': accel_bias, 'gyro_bias': gyro_bias}


def init(mpu):
    # Set ranges to default values
    mpu.set_accel_range(mpu.ACCEL_RANGE_2G)
    mpu.set_gyro_range(mpu.GYRO_RANGE_250DEG)

    # Set the low pass filter to 5
    mpu.set_filter_range(mpu.FILTER_BW_5)

    # Disable temp since not in use
    mpu.disable_temp_sensor()


def read_and_write_mpu6050_data(mpu):
    # Calibrate biases for accelerometer and gyroscope
    biases = calibrate_mpu6050(mpu)
    accel_bias = biases['accel_bias']
    gyro_bias = biases['gyro_bias']
    print(f"Accelerometer Bias: {accel_bias}")
    print(f"Gyroscope Bias: {gyro_bias}")

    gyro_range = mpu.read_gyro_range(True)
    accel_range = mpu.read_accel_range(True)

    while True:
        # Get the current accel and gyro data
        accel_data = mpu.get_accel_data(accel_range,g=False)  # Get accelerometer data in m/s²
        gyro_data = mpu.get_gyro_data(gyro_range)

        # Remove the inherent deviation of the gyro
        gyro_data['x'] -= gyro_bias['x']
        gyro_data['y'] -= gyro_bias['y']
        gyro_data['z'] -= gyro_bias['z']

        # Apply accelerometer bias correction
        accel_data['x'] -= accel_bias['x']
        accel_data['y'] -= accel_bias['y']
        accel_data['z'] -= accel_bias['z']

        # Print or write the data (currently printing to console)
        print(f"Accel Data: {accel_data}, Gyro Data: {gyro_data}")

        # Wait for 1 second
        time.sleep(1)


def calculate_angles(mpu, biases):
    # Initialize variables for angles
    pitch = 0.0
    roll = 0.0
    DT = 0.02
    AA = 0.98
    gyroXangle = 0.0
    gyroYangle = 0.0
    gyroZangle = 0.0
    CFangleX = 0.0
    CFangleY = 0.0

    accel_bias = biases['accel_bias']
    gyro_bias = biases['gyro_bias']
    cnt = 0

    gyro_range = mpu.read_gyro_range(True)
    accel_range = mpu.read_accel_range(True)
    # Main loop to continuously update angles
    while True:
        cnt += 1
        t1 = time.time()
        # Get accelerometer and gyroscope data
        accel_data = mpu.get_accel_data(accel_range,g=False)  # Get accelerometer data in m/s²
        gyro_data = mpu.get_gyro_data(gyro_range)

        # Remove the inherent deviation of the gyro
        gyro_data['x'] -= gyro_bias['x']
        gyro_data['y'] -= gyro_bias['y']
        gyro_data['z'] -= gyro_bias['z']

        # Apply accelerometer bias correction
        accel_data['x'] -= accel_bias['x']
        accel_data['y'] -= accel_bias['y']
        accel_data['z'] -= accel_bias['z']

        # Calculate the angles from the gyro
        gyroXangle += gyro_data['x'] * DT
        gyroYangle += gyro_data['y'] * DT
        gyroZangle += gyro_data['z'] * DT

        # Convert Accelerometer values to degrees
        AccXangle = math.degrees(math.atan2(accel_data['x'], math.sqrt(accel_data['y'] ** 2 + accel_data['z'] ** 2)))
        AccYangle = math.degrees(math.atan2(accel_data['y'], accel_data['z']))

        # Apply the complementary filter
        CFangleX = AA * (CFangleX + gyro_data['x'] * DT) + AccXangle - AA * AccXangle
        CFangleY = AA * (CFangleY + gyro_data['y'] * DT) + AccYangle - AA * AccYangle

        if cnt == 10:
            # Print or use the angles
            print(f"Pitch (CFangleX): {CFangleX:.2f} degrees, Roll (CFangleY): {CFangleY:.2f} degrees")
            cnt = 0

        t2 = time.time()
        time.sleep(max(0, DT - t2 + t1))


def marko_ideja(mpu, biases):
    # Initialize variables for angles
    pitch = 0.0
    roll = 0.0
    DT = 0.02
    AA = 0.98
    gyroXangle = 0.0
    gyroYangle = 0.0
    gyroZangle = 0.0
    CFangleX = 0.0
    CFangleY = 0.0

    accel_bias = biases['accel_bias']
    gyro_bias = biases['gyro_bias']
    cnt = 0
    # Main loop to continuously update angles
    while True:
        cnt += 1
        t1 = time.time()
        # Get accelerometer and gyroscope data
        accel_data = mpu.get_accel_data(g=False)  # Get accelerometer data in m/s²
        gyro_data = mpu.get_gyro_data()

        # Remove the inherent deviation of the gyro
        gyro_data['x'] -= gyro_bias['x']
        gyro_data['y'] -= gyro_bias['y']
        gyro_data['z'] -= gyro_bias['z']

        # Apply accelerometer bias correction
        accel_data['x'] -= accel_bias['x']
        accel_data['y'] -= accel_bias['y']
        accel_data['z'] -= accel_bias['z']

        f = math.sqrt(accel_data['y'] ** 2 + accel_data['z'] ** 2 + accel_data['y'] ** 2)
        alfa = math.degrees(math.acos(max(min(f / mpu.GRAVITIY_MS2, 1), -1)))

        if cnt == 10:
            # Print or use the angles
            print(f"Angle:  {alfa:.2f} degrees")
            cnt = 0

        t2 = time.time()
        time.sleep(max(0, DT - t2 + t1))


# Example usage:
mpu = mpu6050(0x68)
mpu.reset_mpu6050()
init(mpu)
time.sleep(1)
bias = calibrate_mpu6050(mpu)
print(bias)
calculate_angles(mpu, bias)